concert/ext/tangoservers/qa.py

```diff
@@ -0,0 +1,318 @@
+        #lpf = MedianFilter(kernel_size=3)
+        # Process all projections for potential marker centroids but run curve-fit conditionally.
+        async for projection in lpf(ffc(producer)):
+            thresh: ArrayLike = skf.threshold_multiotsu(projection, classes=num_intensity_classes)
+            mask: ArrayLike = projection < thresh.min()
+            labels: ArrayLike = sms.label(mask)
+            regions: List[RegionProperties] = sms.regionprops(label_image=labels)
+            regions = sorted(regions, key=lambda region: region.area, reverse=True)
+            # TODO: Here we implicitly assume that the detected number of regions would always be
+            # greater than the expected number of markers. In practice, this is an edge case because
+            # it might not always be true.
+            if len(regions) > num_markers:
+                regions = regions[:num_markers]
+            marker_centroids.append(np.array(
+                list(map(lambda region: region.centroid, regions))))
+            # We start curve-fit after initial wait window is passed. This is required for the
+            # optimization method to be able to yield some values for the parameters. This wait
+            # window is experimental and we take a conservative approach to choose one to avoid
+            # RuntimeError during curve-fit.
+            if projection_count < wait_window:
+                self.info_stream(
+                    "%s: Skipping estimation until [%d/%d]",
+                    self.__class__.__name__,
+                    projection_count + 1,
+                    wait_window
+                )
+            else:
+                # We start curve-fit after wait window is passed, however we do that with a
+                # configurable offset of projections to gain momentum.
+                if projection_count % self._estm_offset == 0:
+                    try:
+                        rot_axes:  List[np.float64] = []
+                        # TODO: This is a potential edge case. Can this be guaranteed that the
+                        # segmentation algorithm would always be able to detect desired number of
+                        # markers and give us their centroids ?
+                        # OR, a better idea could be that we infer the detected number of markers
+                        # from the shape of the accumulation array. That actually could introduce
+                        # errors in estimation. This needs to be explored and experimented with.
+                        # Ideal solution is the simplest one that works reasonably well all the
+                        # time and not perfectly only some times.
+                        for mix in range(num_markers):
+                            x: ArrayLike = self._angle_dist[:projection_count + 1]
+                            y: ArrayLike = np.array(marker_centroids)[:, mix, 1]
+                            params, _ = sop.curve_fit(f=self.opt_func,
```

Comment:
How long does this take compared to pre-processing?

Reply:
I mean how much does the flat correction, blurring, multi-otsu and curve_fit take? What is the bottleneck right now?

Reply:
With the new implementation we removed multi_otsu and blurring. Assuming other parts of the algorithm are working only thing to focus on now is the curve_fit.